amethyst_derive/src/lib.rs

```diff
@@ -0,0 +1,93 @@
+#![recursion_limit = "256"]
+
+extern crate amethyst_core;
+extern crate proc_macro;
+extern crate proc_macro2;
+#[macro_use]
+extern crate syn;
+#[macro_use]
+extern crate quote;
+
+use proc_macro::TokenStream;
+use syn::{DeriveInput, Ident, Meta, NestedMeta, Data, Type};
+
+#[proc_macro_derive(EventReader, attributes(reader))]
+pub fn event_reader_derive(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+    let gen = impl_event_reader(&ast);
+    gen.into()
+}
+
+fn impl_event_reader(ast: &DeriveInput) -> proc_macro2::TokenStream {
+    let event_name = &ast.ident;
+
+    let mut reader_name : Option<Ident> = None;
+    for meta in ast.attrs.iter()
+        .filter(|attr| attr.path.segments[0].ident == "reader")
+        .map(|attr| attr.interpret_meta().expect("reader attribute incorrectly defined")) {
+        match meta {
+            Meta::List(l) => {
+                for nested_meta in l.nested.iter() {
+                    match *nested_meta {
+                        NestedMeta::Meta(Meta::Word(ref word)) => {
+                            reader_name = Some(word.clone());
+                        },
+                        _ => panic!("reader attribute does not contain a single name"),
+                    }
+                }
+            },
+            _ => (),
+        };
+    }
+    let reader_name = reader_name.expect("No reader type given");
+
+    let tys = collect_field_types(&ast.data);
+    let tys = &tys;
+
+    let reads : Vec<_> = (0..tys.len()).map(|n| {
+        quote! {
+            events.extend(data.#n.read(self.#n.as_mut().expect("ReaderId undefined, has setup been run?")).cloned().map(|e| #event_name::from(e)));
+        }
+    }).collect();
+    let setups : Vec<_> = (0..tys.len()).map(|n| {
+        let ty = &tys[n];
+        quote! {
+            self.#n = Some(res.fetch_mut::<_core::shrev::EventChannel<#ty>>().register_reader());
+        }
+    }).collect();
+    quote! {
+        extern crate amethyst_core as _core;
+
+        #[allow(missing_docs)]
+        #[derive(Default)]
+        pub struct #reader_name(
```

Comment:
If its not present, could you generate it as #derived_nameReader ?

Reply:
Well, i could, but how do we document that behaviour? To use it you need to feed the EventReader implemetation into CoreApplication, and if we create some type name internally how will the user know?

Reply:
I feel it's better to force the user to supply this type.

Reply:
Added a very explicit error message.